crates/modelardb_server/src/storage/data_transfer.rs

```diff
@@ -93,18 +91,24 @@ impl DataTransfer {
         };
 
         // Record the initial used disk space.
-        let initial_disk_space: usize = compressed_files.values().sum();
+        let initial_disk_space: usize = compressed_files.iter().map(|kv| *kv.value()).sum();
+
+        // unwrap() is safe as lock() only returns an error if the lock is poisoned.
         data_transfer
             .used_disk_space_metric
-            .write()
-            .await
+            .lock()
+            .unwrap()
             .append(initial_disk_space as isize, true);
 
         // Check if data should be transferred immediately.
-        for ((table_name, column_index), size_in_bytes) in compressed_files.iter() {
+        for table_name_column_index_size_in_bytes in compressed_files.iter() {
```

Comment:
I agree that automatically deconstructing the keys and values would be nice, unfortunately when [accessing elements](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html#method.get) from a [`DashMap`](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html#) a [Ref](https://docs.rs/dashmap/latest/dashmap/mapref/one/struct.Ref.html) struct with private fields is returned instead of the tuple.

Reply:
That is odd, I assume you mean that this causes some run-time issue because I can make the suggested change and compile without issues. The following code compiles for me:

```rust
for ((table_name, column_index), size_in_bytes) in compressed_files
    .into_iter()
    .collect::<Vec<((String, u16), usize)>>()
{
    if size_in_bytes >= transfer_batch_size_in_bytes {
        data_transfer
            .transfer_data(&table_name, column_index)
            .await
            .map_err(|err| IOError::new(Other, err.to_string()))?;
    }
}
```

Maybe try again but If this is a run-time issue then you can just resolve this.

Reply:
Yes it is possible to change the code as you suggest, but then all of the data in `compressed_files` is copied into a new `Vec` which seems unnecessary for such a small change. So my previous comment was about matching the output `DashMap.iter()`.